pkg/storage/consul/consul_helper.go

```diff
@@ -0,0 +1,706 @@
+
+	// TODO: record metrics
+	if err != nil {
+		return 0, toStorageErr(err, key, 0)
+	}
+
+	// unlike etcd, reads are not rafted, so they don't get an index of their own
+	// so in order to version the resulting list consistantly, we apply the index
+	// of the most recent member
+	maxIndex := uint64(0)
+	for _, item := range kvPairs {
+		if item.ModifyIndex > maxIndex {
+			maxIndex = item.ModifyIndex
+		}
+	}
+
+	err = h.decodeKVPairList(kvPairs, filter, listPtr)
+	if err != nil {
+		return 0, err
+	}
+
+	if err := h.versioner.UpdateList(listObj, maxIndex); err != nil {
+		return 0, err
+	}
+
+	return maxIndex, nil
+}
+
+func (h *consulHelper) decodeKVPairList(kvPairs []*consulapi.KVPair, filter storage.FilterFunc, slicePtr interface{}) error {
+	trace := util.NewTrace("decodeNodeList " + getTypeName(slicePtr))
+	defer trace.LogIfLong(400 * time.Millisecond)
+	v, err := conversion.EnforcePtr(slicePtr)
+	if err != nil || v.Kind() != reflect.Slice {
+		// This should not happen at runtime.
+		panic("need ptr to slice")
+	}
+
+	trace.Step("Decoding KVPairs")
+	for _, kv := range kvPairs {
+		trace.Step("Decoding key " + kv.Key + " START")
+		obj, _, err := h.codec.Decode(kv.Value, nil, reflect.New(v.Type().Elem()).Interface().(runtime.Object))
+		trace.Step("Decoding key " + kv.Key + " END")
+		if err != nil {
+			return err
+		}
+
+		// being unable to set the version does not prevent the object from being extracted
+		_ = h.versioner.UpdateObject(obj, kv.ModifyIndex)
+		if filter(obj) {
+			v.Set(reflect.Append(v, reflect.ValueOf(obj).Elem()))
+		}
+	}
+	trace.Step("Decoded KVPairs")
+
+	return nil
+}
+
+func (h *consulHelper) transformKeyName(keyIn string) string {
+	return strings.Trim(keyIn, "/")
+}
+
+func (h *consulHelper) archiveEvent(key string, objType reflect.Type, eventType watch.EventType) {
+	objOut := reflect.New(objType).Interface().(runtime.Object)
+	err := h.Get(context.Background(), key, objOut, false)
+	if err != nil {
+		//TODO: no panic
+		log.Panicf("Failed to get %s", err)
+	}
+
+	key = h.prefixConsulKey(key)
+	key = h.transformKeyName(key)
+
+	data, err := runtime.Encode(h.codec, objOut)
+	if err != nil {
+		panic("Failed ot encode object")
+	}
+
+	rv := objOut.(meta.Object).GetResourceVersion()
+
+	log.Printf("Archived %s with %s at %s", key, string(eventType), rv)
+	//wrap that into a dummy object and encode globally
+	dummy := ConsulEvent{
+		Object: data,
+		Type:   string(eventType),
+	}
+	dummyEnc, err := ToGOB64(dummy)
+	if err != nil {
+		utilruntime.HandleError(err)
+		//continue or stop then?
+	}
+
+	eventKey := fmt.Sprintf("%s/%s/%s", h.eventHistoryPrefix, key, rv)
+	log.Printf("Key: %s", key)
+	log.Printf("EventKey: %s", eventKey)
+
+	//create KVPair to store it
+	pair := &consulapi.KVPair{
+		Key:   eventKey,
+		Value: []byte(dummyEnc),
+	}
+
+	//store the event with it's version for possible later needs
+	_, err = h.consulKv.Put(pair, nil)
+	if err != nil {
+		//TODO: don't panic
```

Comment:
Should I introduce a new error to `storage.Errors` for this case or reuse an existing one like `ErrCodeResourceVersionConflicts` or `ErrCodeUnreachable`? Both are intended for different cases I guess.
Something like `ErrCodeFailedStoreEventHistory` could be introduced